Fix imports and add isort

joholl opened the pull request:
We have wildcard imports (from ... import *`). That is bad practice.
Additionally, the imports are not sorted alphabetically, currently.

Fix imports and add isort to linters. I also added a linters script to call before a commit for convenience:

```bash
$ .ci/linters.sh
```

Eventually, I'd like to add `pylint`, a de-facto standard and maybe `mypy` and `pydocstyle`.

williamcroberts reviewed (dismissed):
Generally I try and leave large sweeping formatting changes till the end, so folks aren't battling  oodles of conflicts over more productive tasks. @whooo are you good with these now? I'm fine with them, they merge cleanly in my tree.

williamcroberts commented:
tests are failing, not sure why.

joholl commented:
Oh ok, I don't want to produce annoying conflicts. This is an attempt to battle technical dept and one step on the way to fix #163 eventually (because we need a linter for that and at least pylint floods stderr with import issues).

If you prefer that, I'll postpone big linter fixes.

whooo commented:
I don't generally have anything against this, but when it comes to types.py and the tests I rather see a wildcard import, at least for test_esapi.py, as the import list will probably continue to grow and being able to easily throw a test together is good, at least for my workflow :-)

I think keeping __init__.py would be nice as well, doing something like ```from tpm2_pytss import ESAPI, TPM2B_DIGEST``` is useful. One thing I see the python bindings being really good for is easy prototyping and just testing something out so just being able to do ```from tpm2_pytss import *``` and then continue with your code would provide a good experience

whooo commented on test/test_esapi.py:
```diff
@@ -3,9 +3,83 @@
+    TPM2B_ATTEST,
+    TPM2B_AUTH,
+    TPM2B_CREATION_DATA,
+    TPM2B_DATA,
+    TPM2B_DIGEST,
+    TPM2B_ECC_PARAMETER,
+    TPM2B_ECC_POINT,
+    TPM2B_ENCRYPTED_SECRET,
+    TPM2B_EVENT,
+    TPM2B_ID_OBJECT,
+    TPM2B_IV,
+    TPM2B_MAX_BUFFER,
+    TPM2B_NAME,
+    TPM2B_NONCE,
+    TPM2B_NV_PUBLIC,
+    TPM2B_PRIVATE,
+    TPM2B_PUBLIC,
+    TPM2B_PUBLIC_KEY_RSA,
+    TPM2B_SENSITIVE,
+    TPM2B_SENSITIVE_CREATE,
+    TPM2B_SENSITIVE_DATA,
+    TPM2B_TEMPLATE,
+    TPMA_OBJECT,
+    TPML_ALG,
+    TPML_CC,
+    TPML_DIGEST_VALUES,
+    TPML_PCR_SELECTION,
+    TPMS_ALGORITHM_DETAIL_ECC,
+    TPMT_HA,
+    TPMT_PUBLIC,
+    TPMT_PUBLIC_PARMS,
+    TPMT_RSA_DECRYPT,
+    TPMT_SIG_SCHEME,
+    TPMT_SIGNATURE,
+    TPMT_SYM_DEF,
+    TPMT_SYM_DEF_OBJECT,
+    TPMT_TK_CREATION,
+    TPMT_TK_HASHCHECK,
+    TPMT_TK_VERIFIED,
+)
```
These should be imported from tpm2_pytss.types

joholl replied:
You're right, my bad

whooo commented on test/test_types.py:
```diff
@@ -6,7 +6,38 @@
+    TPMT_HA,
+    TPMT_PUBLIC,
+    TPMU_HA,
+    ffi,
```
ffi should come from _libtpm2_pytss

joholl replied:
Yup

williamcroberts commented:
Ahh I just noticed that __init__.py is getting dropped. Yeah we want to keep that IIUC, but this area of python is my least knowledgeable area. At the end of the day, I want the most simple example to be `from tpm_pytss import *`, and it just works. While import * is frowned upon, once an import list gets really large, I think its simpler then to have to keep adding every little thing and minimizes the size of the python file. No one wants a a 50+ line import statement.

joholl commented:
Yes, I threw out all asterisk imports in the pytss code. For rapid coding we (and the user) can still use it, though. Removing the imports from `__init__.py` just means that you have to import the names via dot notation:

```python
from tpm2_pytss.types import *
from tpm2_pytss.ESAPI import *
# ...
```

I can see that the imports in our own code are long-ish now (on my side, my pycharm hides them for me). I could easily change the formatting so that we do not have one import per line.

I understand that is can be slightly more effort to use the dot notation here, but I feel it adds clarity and it seems to be the python way to do things. There are probably good reasons but, to be frank, I am not the most experienced python expert either ;)

whooo commented:
> Yes, I threw out all asterisk imports in the pytss code. For rapid coding we (and the user) can still use it, though. Removing the imports from `__init__.py` just means that you have to import the names via dot notation:
> 
> ```python
> from tpm2_pytss.types import *
> from tpm2_pytss.ESAPI import *
> # ...
> ```
> 
Yes, but that easily grows and requires an understanding of where everything is located, one example from this module is the Wrap function, it's not obvious where it's located without looking at the code. An external example is the cryptography module which would be hard to use without the extensive documentation.
 
> I understand that is can be slightly more effort to use the dot notation here, but I feel it adds clarity and it seems to be the python way to do things. There are probably good reasons but, to be frank, I am not the most experienced python expert either ;)
I agree on the point of clarity, but they aren't exclusive to each other and from an outside perspective it might not be easy to understand what to import.

I don't want to keep the rest of this PR from being merged so my suggestion would be so revisit the removal of ```__init__.py``` later on, after we have improved the documentation, but before the first release candidate for the new bindings.
By being strict about full path imports in the documentation we can encourage that usage as well.

williamcroberts commented:
> Yes, I threw out all asterisk imports in the pytss code. For rapid coding we (and the user) can still use it, though. Removing the imports from `__init__.py` just means that you have to import the names via dot notation:
> 
> ```python
> from tpm2_pytss.types import *
> from tpm2_pytss.ESAPI import *
> # ...
> ```
> 
> I can see that the imports in our own code are long-ish now (on my side, my pycharm hides them for me). I could easily change the formatting so that we do not have one import per line.
> 
> I understand that is can be slightly more effort to use the dot notation here, but I feel it adds clarity and it seems to be the python way to do things. There are probably good reasons but, to be frank, I am not the most experienced python expert either ;)

IIUC, it means that to use ESAPI I need to import ESAPI and types, which I feel is awkward. I just want to import the API and use it.

williamcroberts commented on setup.cfg:
```diff
@@ -43,6 +43,7 @@ dev =
     twine
     setuptools_scm[toml]>=3.4.3
     black==19.10b0
+    isort==5.9.1
```
any reason we're pinning isort?

joholl replied:
I just pinned it because black was pinned, as well. Besides that there is no particular reason. Do you want me to unpin it?

williamcroberts replied:
black is pinned for some bug that makes the formatting different and will cause issues if it's not that version. However, could probably be changed as well at some point to something `>=` once we identify the version thats borked.

williamcroberts replied:
black is also only in beta-4, so it's probably best to pin it until its stable.

williamcroberts replied:
unpin isort

joholl commented:
> Yes, but that easily grows and requires an understanding of where everything is located, one example from this module is the Wrap function, it's not obvious where it's located without looking at the code. An external example is the cryptography module which would be hard to use without the extensive documentation.
 
Agreed. In the end, we need documentation for the API, anyway.

> I don't want to keep the rest of this PR from being merged so my suggestion would be so revisit the removal of __init__.py later on, after we have improved the documentation, but before the first release candidate for the new bindings.

I also agree. We probably don't want to advertise the wildcard imports in the documentation, but that's a conversation for another day.

I re-added the imports back to the `__init__.py` file. At the same time, I suggest using explicit imports (as opposed to wildcard imports) internally. This avoids naming conflicts (for instance `get_ptr()` is defined in `ESAPI` and `FAPI`).

williamcroberts commented on test/test_esapi.py:
```diff
@@ -3,9 +3,81 @@
+    TPMU_HA,
+    TPMU_SYM_KEY_BITS,
+    TPMU_SYM_MODE,
+)
```
Is this really better than `import *`. Why not just use the qualified names?
```python3
from tpm2_pytss import TCTI as TCTI, TSS2_Exception as TSS2_Exception, types
```

joholl replied:
I find using qualified names unnecessary long, but I do not have a strong opinion about this. Having `from tpm2_pytss.types import TPMU_SYM_KEY_BITS` is the default setting e.g. in `black`, `isort` and pycharm. It's just the way it's commonly done because it is non-ambiguous and reasonably short, I guess.

If you want to do it like `import tpm2_pytss.types` and `types.TPMU_SYM_KEY_BITS` every time, instead, that would be fine by me.

williamcroberts replied:
> I find using qualified names unnecessary long, but I do not have a strong opinion about this. Having `from tpm2_pytss.types import TPMU_SYM_KEY_BITS` is the default setting e.g. in `black`, `isort` and pycharm. It's just the way it's commonly done because it is non-ambiguous and reasonably short, I guess.

Those are fine, and since they don't have a zillion classes, aren't overly terrible to deal with.

> 
> If you want to do it like `import tpm2_pytss.types` and `types.TPMU_SYM_KEY_BITS` every time, instead, that would be fine by me.

I'd rather that than an import list that's the majority of the file.

joholl replied:
Ok, as it turns out I made a promise too fast: [black cannot format multiple imports per line](https://github.com/psf/black/issues/204).

To sum this up: either it's a long import list in the tests or it is a wildcard import. I prefer the former because it is unambiguous and best practice.

If you want, I can re-add the wildcard imports (but keep isort for import formatting, i.e. alphabetic sorting etc.).

williamcroberts replied:
> Ok, as it turns out I made a promise too fast: [black cannot format multiple imports per line](https://github.com/psf/black/issues/204).
> 
> To sum this up: either it's a long import list in the tests or it is a wildcard import. I prefer the former because it is unambiguous and best practice.

Yeah but sometimes it just doesn't make sense in certain situations. I think this is one of those situations. Perhaps a rule or guideline here where if the import list is over 10 things long or something like that....

Generally the reason you avoid wildcard imports are things like [this](https://rules.sonarsource.com/python/RSPEC-2208):
```
It could lead to conflicts between names defined locally and the ones imported.
It reduces code readability as developers will have a hard time knowing where names come from.
It clutters the local namespace, which makes debugging more difficult.
```

For item 1, the TPM isn't really usable without the types, you need em all.
For item 2, a 100 line import I think messes with readability more, and with IDE's the resolution of the class name is easily understood.
For item 3, yep it does, but see 1.

> 
> However, if you want, I can re-add the wildcard imports (but keep isort for import formatting, i.e. alphabetic sorting etc.).

Please do that. We can revisit explicit import lists later.

williamcroberts commented on .ci/run.sh:
```diff
@@ -66,10 +66,6 @@ function run_whitespace() {
   fi
 }
 
-function run_style() {
```
Is their any reason style doesn't do both things? We also end up spinning an extra VM instance on GitHub which is expensive and takes more time then just running all are checks in one thing. We can change the logic to not fail fast, so we get everything that's wrong over the fist set of errors from the first tool being executed.

joholl replied:
Unlike other linters, `isort` runs really quickly, yes. Style checking is not a bottle neck either way, so it doesn't really matter performance-wise. I find it more clean and organized to have one job per linter, especially if we want to add more linters in the future. That way you see instantly which linter failed and can instantly see the relevant log (instead of searching for the relevant bits).

williamcroberts replied:
Its the VM spin-up time, nothing to do with whats executed. Also, we can only launch so many VMs at a time (offhand I think its 4). I'd rather dedicate parallel VM's to running actual function checks over additional "linter" checks. At some point in time, we will run multiple Python version checks before 1.0.

Right now we will have 3 nodes running just for style checks and if/when we add the linter, there's another instance.

What you can do is make each check it's own job, but run it in one node as multiple jobs. I think you can just mark the jobs as continue-on-error: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error

joholl replied:
Oh, that's a very good point! `continue-on-error` shows the stage as passed even though it failed, but I found another solution which continues on fail using `if: always()`.

williamcroberts commented on test/test_esapi.py:
```diff
@@ -3804,7 +3876,3 @@ def test_PolicyTicket(self):
-
-
-if __name__ == "__main__":
-    unittest.main()
```
I like the ability to run these as standalone scripts.

williamcroberts commented on test/test_types.py:
```diff
@@ -6,7 +6,41 @@
 import itertools
 import unittest
 
-from tpm2_pytss import *
```
keep this one as import * as well. The shorter lists can survive.

williamcroberts commented on test/test_types.py:
```diff
@@ -1297,7 +1331,3 @@ def test_TPML_DIGEST(self):
-
-
-if __name__ == "__main__":
-    unittest.main()
```
keep this

williamcroberts commented on tpm2_pytss/ESAPI.py:
```diff
@@ -2,12 +2,66 @@
 
-from ._libtpm2_pytss import lib
-
-from .types import *
```
keep this wildcard one as well

williamcroberts commented on tpm2_pytss/tsskey.py:
```diff
@@ -3,11 +3,33 @@
 
 import warnings
-from ._libtpm2_pytss import lib
-from .types import *
```
keep this one...

joholl commented:
Thanks for the review! I added the requested changes and rebased:

* Removed the big wildcard imports
* Unpinned isort
* Re-added `unittest.main()` at the end of tests (and added it to every test script)

williamcroberts commented:
@joholl if you want to get this into 1.0, can you rebase this. We're in the process of planning the release.

williamcroberts commented:
closing
